File: grind75/go/merge-two-sorted-lists.go
package main

import (
	"fmt"
)

type ListNode struct {
	Val  int
	Next *ListNode
}

func main() {

	result := mergeTwoLists(
		&ListNode{Val: 1, Next: &ListNode{Val: 2, Next: &ListNode{Val: 4}}},
		&ListNode{Val: 1, Next: &ListNode{Val: 3, Next: &ListNode{Val: 4}}})

	fmt.Println(result)
}

func mergeTwoLists(list1 *ListNode, list2 *ListNode) *ListNode {
	var result *ListNode
	var current *ListNode

	for list1 != nil && list2 != nil {
		if list1.Val < list2.Val {
			if result == nil {
				result = &ListNode{Val: list1.Val}
				current = result
			} else {
				current.Next = &ListNode{Val: list1.Val}
				current = current.Next
			}
			list1 = list1.Next
		} else {
			if result == nil {
				result = &ListNode{Val: list2.Val}
				current = result
			} else {
				current.Next = &ListNode{Val: list2.Val}
				current = current.Next
			}
			list2 = list2.Next
		}
	}

	for list1 != nil {
		if result == nil {
			result = &ListNode{Val: list1.Val}
			current = result
		} else {
			current.Next = &ListNode{Val: list1.Val}
			current = current.Next
		}
		list1 = list1.Next
	}

	for list2 != nil {
		if result == nil {
			result = &ListNode{Val: list2.Val}
			current = result
		} else {
			current.Next = &ListNode{Val: list2.Val}
			current = current.Next
		}
		list2 = list2.Next
	}

	return result

}

File: grind75/rust/merge-two-sorted-lists.rs
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}
fn main() {
    pub fn merge_two_lists(
        mut list1: Option<Box<ListNode>>,
        mut list2: Option<Box<ListNode>>,
    ) -> Option<Box<ListNode>> {
        let mut result = Some(Box::new(ListNode::new(0)));
        let mut current = &mut result;

        while list1.is_some() && list2.is_some() {
            let val1 = list1.as_ref().unwrap().val;
            let val2 = list2.as_ref().unwrap().val;

            if val1 < val2 {
                let next = list1.as_mut().unwrap().next.take();
                current.as_mut().unwrap().next = list1;
                list1 = next;
            } else {
                let next = list2.as_mut().unwrap().next.take();
                current.as_mut().unwrap().next = list2;
                list2 = next;
            }

            current = &mut current.as_mut().unwrap().next;
        }
        if list1.is_some() {
            current.as_mut().unwrap().next = list1;
        } else {
            current.as_mut().unwrap().next = list2;
        }

        return result.unwrap().next;
    }

    let list1 = Some(Box::new(ListNode {
        val: 1,
        next: Some(Box::new(ListNode { val: 2, next: None })),
    }));

    let list2 = Some(Box::new(ListNode {
        val: 1,
        next: Some(Box::new(ListNode { val: 3, next: None })),
    }));

    let result = merge_two_lists(list1, list2);

    println!("{:?}", result);
}
